logger_cb.go

```diff
@@ -0,0 +1,157 @@
+package libbpfgo
+
+/*
+#include <bpf/libbpf.h>
+*/
+import "C"
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+)
+
+// This callback definition needs to be in a different file from where it is declared in C
+// Otherwise, multiple definition compilation error will occur
+
+// loggerCallback is called by libbpf_print_fn() which in turn is called by libbpf
+//
+//export loggerCallback
+func loggerCallback(libbpfPrintLevel int, libbpfOutput *C.char) {
+	var (
+		level    int
+		goOutput string
+	)
+
+	goOutput = C.GoString(libbpfOutput)
+	goOutput = strings.TrimSuffix(goOutput, "\n")
+
+	for _, fnFilterOut := range callbacks.LogFilters {
+		if fnFilterOut != nil {
+			if fnFilterOut(libbpfPrintLevel, goOutput) {
+				return
+			}
+		}
+	}
+
+	callbacks.Log(level, goOutput)
+}
+
+const (
+	LibbpfWarnLevel  = int(C.LIBBPF_WARN)
+	LibbpfInfoLevel  = int(C.LIBBPF_INFO)
+	LibbpfDebugLevel = int(C.LIBBPF_DEBUG)
+)
+
+type Callbacks struct {
+	Log        func(level int, msg string, keyValues ...interface{})
+	LogFilters []func(libLevel int, msg string) bool
+}
+
+var callbacks = Callbacks{
+	Log: logFallback,
+	LogFilters: []func(libLevel int, msg string) bool{
+		LogFilterLevel,
+		LogFilterOutput,
+	},
+}
+
+// SetLoggerCbs receives Callbacks type to be used to log libbpf outputs and to filter out those outputs
+func SetLoggerCbs(cbs Callbacks) {
+	if cbs.Log == nil {
+		cbs.Log = logFallback
+	}
+
+	callbacks = cbs
+}
+
+// logFallback:
+// - level is ignored in this stage
+// - type coercion only takes care of string types
+// - keyValues is not required to contain pairs
+// - outputs all to stderr
+func logFallback(level int, msg string, keyValues ...interface{}) {
+	var (
+		args   = make([]string, 0)
+		outMsg = msg
+	)
+
+	for _, v := range keyValues {
+		if s, ok := v.(string); ok {
+			outMsg += " [%s]"
+			args = append(args, s)
+		}
+	}
+
+	outMsg += "\n"
+	if len(keyValues) > 0 {
+		fmt.Fprintf(os.Stderr, outMsg, args)
+	} else {
+		fmt.Fprint(os.Stderr, outMsg)
+	}
+}
+
+// LogFilterLevel filters by checking its print level
+// In case the consumer defines its own filters functions via SetLoggerCbs, this can also be passed
+func LogFilterLevel(libbpfPrintLevel int, output string) bool {
+	return libbpfPrintLevel != LibbpfWarnLevel
+}
+
+// LogFilterOutput filters out some errors by using regex
+// In case the consumer defines its own filters functions via SetLoggerCbs, this can also be passed
+func LogFilterOutput(libbpfPrintLevel int, output string) bool {
+	var (
+		matched bool
+		err     error
+	)
+
+	// BUG: https:/github.com/aquasecurity/tracee/issues/1676
+
+	// triggered by: libbpf/src/nlattr.c->libbpf_nla_dump_errormsg()
+	// "libbpf: Kernel error message: %s\n"
+	// 1. %s = "Exclusivity flag on"
+
+	matched, err = regexp.MatchString(`libbpf:.*Kernel error message:.*Exclusivity flag on`, output)
```

Comment:
Do not use the `Match` functions directly from `regexp`,  as they recompile the regex per call. Please compile the regex outside the function scope once then reuse it.